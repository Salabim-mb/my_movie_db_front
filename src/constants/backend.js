const backend_path = process.env.REACT_APP_BACKEND_PATH;

export const backend = {
    MOVIES: backend_path + "movies/",
    USERS: backend_path + "user/",
    PLAIN: backend_path,

};