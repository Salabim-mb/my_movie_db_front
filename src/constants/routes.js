import MovieList from "../components/MovieList";

const paths = {
    MAIN: "/",
    LOGIN: "/login",
    REGISTER: "/register",
    MOVIES: "/movies",
    NEW_MOVIE: "/movies/new"
};

export default [
    {
        path: paths.MOVIES,
        exact: true,
        component: MovieList
    }r
]

export {paths};