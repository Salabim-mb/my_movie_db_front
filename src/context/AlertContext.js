import React, {useState, createContext} from 'react';

export const AlertContext = createContext({
    open: false,
    message: "",
    setOpen: () => {}
});

export const AlertProvider = (props) => {
    const [open, setOpen] = false;
    const [message, setMessage] = useState("");
    const [variant, setVariant] = useState("success");

    const data = {
        open,
        message,
        variant,
        changeVisibility: (newOpen) => setOpen(newOpen),
        changeMessage: (newMessage, newVariant = "success") => {
            setOpen(false);
            setMessage(newMessage);
            setVariant(newVariant);
        },
        showAlert: (newMessage, newVariant) => {
            data.changeMessage(newMessage, newVariant);
            data.changeVisibility(true);
        },
    };
    return <AlertContext.Provider value={data} {...props} />;
};