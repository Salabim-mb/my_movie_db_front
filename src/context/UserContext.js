import React, {useState, useEffect} from 'react';
import Cookies from "universal-cookie";
import {getHeaders} from "../utils/CORSHeaders";
import {backend} from "../constants/backend";

export const UserContext = React.createContext({
    token: undefined,
    data: undefined,
    login: () => {},
    logout: () => {}
});
const cookies = new Cookies();

const getUserData = async (token) => {
    const url = `${backend.USERS}data/`;
    const response = await fetch(url, {
        method: "GET",
        headers: getHeaders(token),
    });

    if (response.status === 200) {
        return await response.json();
    } else {
        throw response.status;
    }
};

export const UserProvider = (props) => {
    const [token, setToken] = useState(cookies.get("token"));
    const [data, setData] = useState(cookies.get("data"));

    useEffect(() => {
        const fetchUserData = async() => {
            try {
                const res = await getUserData(token);
                setData(res.data);
                cookies.set("data", res.data, {
                    path: "/",
                    expires: setCookieDate()
                });
            } catch(e) {
                console.log(e);
            }
        }
        if (token) {
            fetchUserData();
        }
    }, [token]);

    const user = {
        token,
        data,
        changeToken: (newToken) => {
            cookies.set("token", newToken, {
                path: "/",
                expires: setCookieDate()
            });
            setToken(newToken);
        },
        changeData: (newData) => {
            cookies.set("data", newData, {
                path: "/",
                expires: setCookieDate()
            });
            setData(newData);
        },
        login: (newToken, newData) => {
            user.changeToken(newToken);
            user.changeData(newData);
        },
        logout: () => {
            cookies.remove("token", { path: "/" });
            cookies.remove("data", { path: "/" });
            setToken(undefined);
            setData(undefined);
        }
    };

    return <UserContext.Provider value={user} {...props} />;
};

const setCookieDate = () => {
    let date = new Date();
    date.setDate(date.getDate() + 1);
    return date;
};