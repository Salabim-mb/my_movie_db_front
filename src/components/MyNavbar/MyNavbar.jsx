import React from "react";
import {Button, Nav, Navbar} from "react-bootstrap";
import { IndexLinkContainer } from "react-router-bootstrap";
// Normalnie powinno się używać ścieżek absulotnych (przyjmuje się, że folderem domowym projektu jest src)
// tzn ścieżka powinna wyglądać "constants/routes" zamiast "../..itd",
// ale nie wiem czemu to nie działa, a nie chce mi się grzebać w takich pierdołach xd
import {paths} from "../../constants/routes";
import {useLocation} from "react-router";

// Można wrzucić to w () => { return (content) }, ale poniższy zapis () => () jest równoznaczny,
// jeżeli nie chcemy pchać jakiejś wielkiej logiki, to "return" można pominąć

// przykład "klasy" funkcyjnej (nie ma własnego stanu ani logiki za bardzo, ale zależy nam na prostocie,
// bo będzie na każdej stronie i nie będzie odświeżana :)

// komponenty z wielkiej litery (te importowane wyżej) pochodzą z pakietu react-bootstrap, który ma predefiniowane style,
// a te zaimportowałem wcześniej ;)

const MyNavbar = () => {
    let location = useLocation();

    return (
        <Navbar bg="dark" variant="dark">
            <IndexLinkContainer to={paths.MAIN}>
                <Nav.Link className="mr-5">
                    <Navbar.Brand>
                        MyMovieDB
                    </Navbar.Brand>
                </Nav.Link>
            </IndexLinkContainer>
            <Nav className="mr-auto">
                {/*{ "logo" apki, zamykane w IndexLinkContainer, żeby nie odświeżało po kliknięciu (ale nie jest to jedyna metoda) ;) }*/}
                <IndexLinkContainer to={paths.MOVIES}>
                    <Nav.Link>
                        Movie list
                    </Nav.Link>
                </IndexLinkContainer>
            </Nav>
            {/*{ justify-content-end to bootstrap, wrzucamy go w property className. Dzięki temu nasz komponent będzie wyrównany do prawej }*/}
            {
                location.pathname !== paths.NEW_MOVIE &&
                location.pathname !== paths.EDIT_MOVIE && (
                    <Navbar.Collapse className="justify-content-end">
                        <IndexLinkContainer to={paths.NEW_MOVIE}>
                            <Button variant="success">
                                + Add movie
                            </Button>
                        </IndexLinkContainer>
                    </Navbar.Collapse>
                )
            }
        </Navbar>
    )
};
// ^ średnik na końcu można pominąć -- co prawda będzie podkreślone, ale nic się nie stanie. JavaScript jest super <3

export default MyNavbar;