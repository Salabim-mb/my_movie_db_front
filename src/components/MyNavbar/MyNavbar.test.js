import {render, fireEvent} from '@testing-library/react';
import React from "react";
import {createMemoryHistory} from 'history';
import {MemoryRouter, Router} from "react-router-dom";
import MyNavbar from "./MyNavbar";
import {paths} from "../../constants/routes";
import {UserContext} from "../../context/UserContext";
import {backend} from "../../constants/backend";

const renderWithRouter = (
    ui,
    {
        route = "/non-existent-route",
        history = createMemoryHistory({ initialEntries: [route] })
    } = {}
) => {
    return {
        ...render(
            <Router history={history}>{ui}</Router>
        ),
        history
    }
};

describe("MyNavbar", () => {
    let context;
    let apiFail;
    beforeAll(() => {
        global.fetch = jest.fn().mockImplementation((input, init) => {
            return new Promise((resolve, reject) => {
                if (apiFail) {
                    resolve({status: 500})
                } else {
                    resolve({status: 200, json: () => Promise.resolve("it's ok")})
                }
            })
        });
    });

    beforeEach(() => {
        jest.clearAllMocks();
        context = {
            token: "123",
            data: {
                name: "John",
                surname: "Doe",
                email: "jd@jd.jd"
            }
        }
    });

    it("should match snapshot", () => {
        const {container} = render(
            <MemoryRouter>
                <MyNavbar />
            </MemoryRouter>
        );
        expect(container).toMatchSnapshot();
    });

    it("should redirect to form", () => {
        const {history, getByText} = renderWithRouter(
            <MyNavbar />
        );

        fireEvent.click( getByText("+ Add movie") );
        expect(history.location.pathname).toBe(paths.NEW_MOVIE, {exact: false});
    });

    it("should redirect to main page", () => {
        const {history, getByText} = renderWithRouter(
            <MyNavbar />
        );

        fireEvent.click( getByText("MyMovieDB") );
        expect(history.location.pathname).toBe(paths.MAIN, {exact: false});
    });

    it("should log user out", async () => {
        const {history, getByText} = renderWithRouter(
            <UserContext.Provider value={context}>
                <MyNavbar />
            </UserContext.Provider>
        );
        fireEvent.click(getByText("Log out"));

        await expect(fetch).toHaveBeenCalledWith(`${backend.PLAIN}login/`, {
            method: "DELETE",
            headers: {
                "token": context.token,
                "Content-Type": "application/json"
            }
        });

        expect(context?.token).toBe(undefined);
        expect(history.location.pathname).toBe(paths.MAIN);
    });

    it("should log user out even if api failed", async () => {
        apiFail = true;
        const {history, getByText} = renderWithRouter(
            <UserContext.Provider value={context}>
                <MyNavbar />
            </UserContext.Provider>
        );
        fireEvent.click(getByText("Log out"));

        await expect(fetch).toHaveBeenCalledTimes(1);

        expect(context?.token).toBe(undefined);
        expect(history.location.pathname).toBe(paths.MAIN);
    });


});