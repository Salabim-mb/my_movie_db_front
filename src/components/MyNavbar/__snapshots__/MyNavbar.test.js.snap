// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyNavbar should match snapshot 1`] = `
<div>
  <nav
    class="navbar navbar-expand navbar-dark bg-dark"
  >
    <a
      class="mr-5 active nav-link"
      href="/"
    >
      <span
        class="navbar-brand"
      >
        MyMovieDB
      </span>
    </a>
    <div
      class="mr-auto navbar-nav"
    >
      <a
        class="nav-link"
        data-rb-event-key="/movies"
        href="/movies"
      >
        Movie list
      </a>
      <div />
    </div>
    <div
      class="justify-content-end navbar-collapse collapse"
    >
      <a
        class="btn btn-success"
        href="/movies/new"
      >
        + Add movie
      </a>
      <a
        class="btn btn-light"
        href="/login"
      >
        Log in
      </a>
    </div>
  </nav>
</div>
`;
